### PR TITLE
dhtrunner: recover from network changes

### PR DESCRIPTION
**Pull request by sim590**
Using a list of bootstrap nodes, we recover from network loss by pinging those nodes continuously until connected.

### CODE REVIEW

**Comment by sim590** [comment]
This has yet to be tested.

- [ ] test with practical network disconnections.

File: src/dht.cpp [hunk 1]
```diff
@@ -2787,7 +2789,7 @@ Dht::maintainStorage(InfoHash id, bool force, DoneCallback donecb) {
         if (force || id.xorCmp(nodes6.back()->id, myid) < 0) {
             for (auto &value : local_storage->second.getValues()) {
                 const auto& vt = getType(value.data->type);
-                if (force || value.time + vt.expiration > now + MAX_STORAGE_MAINTENANCE_EXPIRE_TIME) {
+                if (force || value.time + vt.expiration > now + std::chrono::minutes(1)) {
```
Comment (sim590) [inline src/dht.cpp]: Oops. This is unrelated.

- [ ] remove this.

File: src/dhtrunner.cpp [hunk 2]
```diff
@@ -325,6 +328,31 @@ DhtRunner::loop_()
...
+                NodeStatus s4, s6;
+                do {
+                    {
+                        std::lock_guard<std::mutex> lck(bootstrap_mtx);
+                        for (const auto& b : bootstrap_nodes) {
+                            bootstrap(b.first, b.second);
+                        }
+                    }
+                    /* wait and see if we get response from a node. */
+                    std::this_thread::sleep_for(std::chrono::seconds(Request::MAX_ATTEMPT_COUNT*Node::MAX_RESPONSE_TIME));
```
Comment (sim590) [inline src/dhtrunner.cpp]: @aberaud: This is the part where I expect comments... ;)

File: src/dht.cpp [hunk 3]
```diff
@@ -2771,7 +2772,8 @@ Dht::maintainStorage(InfoHash id, bool force, DoneCallback donecb) {
         if (force || id.xorCmp(nodes.back()->id, myid) < 0) {
             for (auto &value : local_storage->second.getValues()) {
                 const auto& vt = getType(value.data->type);
-                if (force || value.time + vt.expiration > now + MAX_STORAGE_MAINTENANCE_EXPIRE_TIME) {
+                /* we only put the value if it's not going to expire soon... 1 minute is fair. */
+                if (force || value.time + vt.expiration > now + std::chrono::minutes(1)) {
```
Comment (aberaud) [inline src/dht.cpp]: TODO

File: src/dhtrunner.cpp [hunk 4]
```diff
@@ -571,18 +580,69 @@ DhtRunner::putEncrypted(const std::string& key, InfoHash to, Value&& value, Done
     putEncrypted(InfoHash::get(key), to, std::forward<Value>(value), cb);
 }
 
+void DhtRunner::tryBootstrapCoutinuously() {
+    if (bootstrap_thread.joinable()) {
+        std::cerr << "Previous bootstrap thread running..." << std::endl;
```
Comment (aberaud) [inline src/dhtrunner.cpp]: DHT_LOG

Comment (sim590) [inline src/dhtrunner.cpp]: I wish, but it is protected in `Dht`.

**Comment by aberaud** [comment]
updated here : https://github.com/savoirfairelinux/opendht/pull/139
